File: javascript/07-switch-factory/test/playlist_curator.test.js
import assert from 'node:assert';
import test from 'node:test';
import { PlaylistCurator } from '../src/playlist_curator.js';

test('filters happy tracks', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', tempo: 130 }, { title: 'B', tempo: 100 }];
  const result = curator.create_playlist('happy', tracks);
  assert.strictEqual(result.length, 1);
  assert.strictEqual(result[0].title, 'A');
});

test('filters sad tracks', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', tempo: 80 }, { title: 'B', tempo: 100 }];
  const result = curator.create_playlist('sad', tracks);
  assert.strictEqual(result.length, 1);
  assert.strictEqual(result[0].title, 'A');
});

test('filters workout tracks', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', tempo: 140, energy: 8 }, { title: 'B', tempo: 140, energy: 5 }];
  const result = curator.create_playlist('workout', tracks);
  assert.strictEqual(result.length, 1);
  assert.strictEqual(result[0].title, 'A');
});

test('filters focus tracks', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', instrumental: true }, { title: 'B', instrumental: false }];
  const result = curator.create_playlist('focus', tracks);
  assert.strictEqual(result.length, 1);
  assert.strictEqual(result[0].title, 'A');
});

test('filters party tracks', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', tempo: 120, danceability: 7 }, { title: 'B', tempo: 100, danceability: 5 }];
  const result = curator.create_playlist('party', tracks);
  assert.strictEqual(result.length, 1);
  assert.strictEqual(result[0].title, 'A');
});

test('caps happy playlist at 20', () => {
  const curator = new PlaylistCurator();
  const tracks = Array(25).fill(null).map((_, i) => ({ title: String(i), tempo: 130 }));
  const result = curator.create_playlist('happy', tracks);
  assert.strictEqual(result.length, 20);
});

test('caps sad playlist at 15', () => {
  const curator = new PlaylistCurator();
  const tracks = Array(20).fill(null).map((_, i) => ({ title: String(i), tempo: 80 }));
  const result = curator.create_playlist('sad', tracks);
  assert.strictEqual(result.length, 15);
});

test('defaults to first 10 for unknown mood', () => {
  const curator = new PlaylistCurator();
  const tracks = Array(20).fill(null).map((_, i) => ({ title: String(i), tempo: 100 }));
  const result = curator.create_playlist('mysterious', tracks);
  assert.strictEqual(result.length, 10);
});

test('sorts happy by descending tempo', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', tempo: 130 }, { title: 'B', tempo: 150 }];
  const result = curator.create_playlist('happy', tracks);
  assert.strictEqual(result[0].title, 'B');
  assert.strictEqual(result[1].title, 'A');
});

test('sorts sad by ascending tempo', () => {
  const curator = new PlaylistCurator();
  const tracks = [{ title: 'A', tempo: 80 }, { title: 'B', tempo: 60 }];
  const result = curator.create_playlist('sad', tracks);
  assert.strictEqual(result[0].title, 'B');
  assert.strictEqual(result[1].title, 'A');
});

File: ruby/07-switch-factory/test/playlist_curator_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../src/playlist_curator'

class PlaylistCuratorTest < Minitest::Test
  def setup
    @curator = PlaylistCurator.new
  end

  def test_filters_happy_tracks
    result = @curator.create_playlist('happy', [{ title: 'A', tempo: 130 }, { title: 'B', tempo: 100 }])
    assert_equal 1, result.length
    assert_equal 'A', result[0][:title]
  end

  def test_filters_sad_tracks
    result = @curator.create_playlist('sad', [{ title: 'A', tempo: 80 }, { title: 'B', tempo: 100 }])
    assert_equal 1, result.length
    assert_equal 'A', result[0][:title]
  end

  def test_filters_workout_tracks
    result = @curator.create_playlist('workout', [{ title: 'A', tempo: 140, energy: 8 }, { title: 'B', tempo: 140, energy: 5 }])
    assert_equal 1, result.length
    assert_equal 'A', result[0][:title]
  end

  def test_filters_focus_tracks
    result = @curator.create_playlist('focus', [{ title: 'A', instrumental: true }, { title: 'B', instrumental: false }])
    assert_equal 1, result.length
    assert_equal 'A', result[0][:title]
  end

  def test_filters_party_tracks
    result = @curator.create_playlist('party', [{ title: 'A', tempo: 120, danceability: 7 }, { title: 'B', tempo: 100, danceability: 5 }])
    assert_equal 1, result.length
    assert_equal 'A', result[0][:title]
  end

  def test_caps_happy_playlist_at_20
    tracks = Array.new(25) { |i| { title: i.to_s, tempo: 130 } }
    result = @curator.create_playlist('happy', tracks)
    assert_equal 20, result.length
  end

  def test_caps_sad_playlist_at_15
    tracks = Array.new(20) { |i| { title: i.to_s, tempo: 80 } }
    result = @curator.create_playlist('sad', tracks)
    assert_equal 15, result.length
  end

  def test_defaults_to_first_10_for_unknown_mood
    tracks = Array.new(20) { |i| { title: i.to_s, tempo: 100 } }
    result = @curator.create_playlist('mysterious', tracks)
    assert_equal 10, result.length
  end

  def test_sorts_happy_by_descending_tempo
    result = @curator.create_playlist('happy', [{ title: 'A', tempo: 130 }, { title: 'B', tempo: 150 }])
    assert_equal 'B', result[0][:title]
    assert_equal 'A', result[1][:title]
  end

  def test_sorts_sad_by_ascending_tempo
    result = @curator.create_playlist('sad', [{ title: 'A', tempo: 80 }, { title: 'B', tempo: 60 }])
    assert_equal 'B', result[0][:title]
    assert_equal 'A', result[1][:title]
  end
end
